From 3e3661341e3c1c1781d3115493dd59ad5d7061ea Mon Sep 17 00:00:00 2001
From: Hyun Kwon <hyun.kwon@xilinx.com>
Date: Fri, 5 Jan 2018 17:40:57 -0800
Subject: [PATCH 04/10] dma: xilinx: dpdma: Don't drop the error desc

This patch comes from:
  https://github.com/Xilinx/linux-xlnx.git

Don't blindly drop the error descriptor as the error may
happen with a correct descriptor. If there's a pending
descriptor, the error descriptor is skipped and the next
descriptor is scheduled.

Signed-off-by: Hyun Kwon <hyun.kwon@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
(cherry picked from commit 7b8a15381bd4d4de7cf08d35b741c583a32154dc)
---
 drivers/dma/xilinx/xilinx_dpdma.c | 38 ++++++++++++++++----------------------
 1 file changed, 16 insertions(+), 22 deletions(-)

diff --git a/drivers/dma/xilinx/xilinx_dpdma.c b/drivers/dma/xilinx/xilinx_dpdma.c
index 0964f19..6a3e185 100644
--- a/drivers/dma/xilinx/xilinx_dpdma.c
+++ b/drivers/dma/xilinx/xilinx_dpdma.c
@@ -1687,31 +1687,25 @@ static void xilinx_dpdma_chan_handle_err(struct xilinx_dpdma_chan *chan)
 	xilinx_dpdma_chan_disable(chan);
 	chan->status = IDLE;
 
-	/* Decide if the current descriptor can be rescheduled */
-	if (chan->active_desc) {
-		switch (chan->active_desc->status) {
-		case ACTIVE:
-		case PREPARED:
-			xilinx_dpdma_chan_free_tx_desc(chan,
-						       chan->submitted_desc);
-			chan->submitted_desc = NULL;
-			xilinx_dpdma_chan_free_tx_desc(chan,
-						       chan->pending_desc);
-			chan->pending_desc = NULL;
-			chan->active_desc->status = ERRORED;
+	if (!chan->active_desc)
+		goto out_unlock;
+
+	xilinx_dpdma_chan_dump_tx_desc(chan, chan->active_desc);
+
+	switch (chan->active_desc->status) {
+	case ERRORED:
+		dev_dbg(dev, "repeated error on desc\n");
+	case ACTIVE:
+	case PREPARED:
+		/* Reschedule if there's no new descriptor */
+		if (!chan->submitted_desc)
 			chan->submitted_desc = chan->active_desc;
-			break;
-		case ERRORED:
-			dev_err(dev, "desc is dropped by unrecoverable err\n");
-			xilinx_dpdma_chan_dump_tx_desc(chan, chan->active_desc);
-			xilinx_dpdma_chan_free_tx_desc(chan, chan->active_desc);
-			break;
-		default:
-			break;
-		}
-		chan->active_desc = NULL;
+		chan->active_desc->status = ERRORED;
+		break;
 	}
+	chan->active_desc = NULL;
 
+out_unlock:
 	spin_unlock_irqrestore(&chan->lock, flags);
 }
 
-- 
1.8.5.2.233.g932f7e4

